### ISSUE TITLE
App::draw signal is issued twice when resizing or changing display (osx)

### ISSUE THREAD

**Issue by simongeilfus** [cinder/Cinder#762]
Is this a desired behavior? In some rare cases where you need to have exactly one App::update for each App::draw it might become an issue, so if this is indeed a desired behavior to avoid flickering or other issues, would it be something we could opt out in the App::Settings?

Here's a small test app to reproduce what I'm talking about :

```c++
#include "cinder/app/App.h"
#include "cinder/app/RendererGl.h"
#include "cinder/gl/gl.h"

using namespace ci;
using namespace ci::app;
using namespace std;

class DrawTwiceApp : public App {
  public:
	void draw() override;
    
    int mLastDrawFrame = 0;
};

void DrawTwiceApp::draw()
{
	gl::clear( Color( 0, 0, 0 ) );
    if( getElapsedFrames() == mLastDrawFrame )
        cout << "Draw issued twice \n \tcurrent: " << getElapsedFrames() << endl
        << "\tprevious: " << mLastDrawFrame << endl;
    mLastDrawFrame = getElapsedFrames();
}

CINDER_APP( DrawTwiceApp, RendererGl )
```
Resizing this app window or putting it on a different display will trigger the cout on my mac.

Let me know if I can be of any help on this. I've been looking for the source of the second draw call but couldn't find it. ( There's a [self draw]; in CinderViewMac::setFrameSize but that's not the one, ... )

**Comment by richardeakin** [comment]
I can only test with resizing, as I don't have two displays atm. Looks like we're getting a `[CinderViewMac drawRect:]` call that is spawning from the window update:

```
  [0] = "0   BasicApp                            0x000000010000b599 cinder::app::PlatformCocoa::stackTrace() + 249"
  [1] = "1   BasicApp                            0x0000000100014073 cinder::stackTrace() + 51"
  [2] = "2   BasicApp                            0x0000000100002208 DrawTwiceApp::draw() + 328"
  [3] = "3   BasicApp                            0x00000001000a48da cinder::app::Window::emitDraw() + 138"
  [4] = "4   BasicApp                            0x00000001000185f5 -[WindowImplBasicCocoa draw] + 101"
  [5] = "5   BasicApp                            0x000000010006c476 -[CinderViewMac drawRect:] + 214"
  [6] = "6   AppKit                              0x00007fff94accb19 -[NSView _drawRect:clip:] + 4238"
  [7] = "7   AppKit                              0x00007fff94acb17a -[NSView _recursiveDisplayAllDirtyWithLockFocus:visRect:] + 1875"
  [8] = "8   AppKit                              0x00007fff94acb57e -[NSView _recursiveDisplayAllDirtyWithLockFocus:visRect:] + 2903"
  [9] = "9   AppKit                              0x00007fff94ac9016 -[NSView _recursiveDisplayRectIfNeededIgnoringOpacity:isVisibleRect:rectIsVisibleRectForView:topView:] + 913"
  [10] = "10  AppKit                              0x00007fff94ac8773 -[NSThemeFrame _recursiveDisplayRectIfNeededIgnoringOpacity:isVisibleRect:rectIsVisibleRectForView:topView:] + 333"
  [11] = "11  AppKit                              0x00007fff94ac52cb -[NSView _displayRectIgnoringOpacity:isVisibleRect:rectIsVisibleRectForView:] + 2761"
  [12] = "12  AppKit                              0x00007fff94aa3e2d -[NSView displayIfNeeded] + 1876"
...
```

in addition the one triggered by `timerFired:`:
```
 [0] = "0   BasicApp                            0x000000010000b599 cinder::app::PlatformCocoa::stackTrace() + 249"
  [1] = "1   BasicApp                            0x0000000100014073 cinder::stackTrace() + 51"
  [2] = "2   BasicApp                            0x0000000100002238 DrawTwiceApp::draw() + 376"
  [3] = "3   BasicApp                            0x00000001000a48da cinder::app::Window::emitDraw() + 138"
  [4] = "4   BasicApp                            0x00000001000185f5 -[WindowImplBasicCocoa draw] + 101"
  [5] = "5   BasicApp                            0x000000010006c476 -[CinderViewMac drawRect:] + 214"
  [6] = "6   BasicApp                            0x000000010006c34c -[CinderViewMac draw] + 188"
  [7] = "7   BasicApp                            0x00000001000153f6 -[AppImplMac timerFired:] + 1142"
...
```

Also, there's a `[self draw]` call in CinderViewMac's `setFrameSize:` method ([here](https://github.com/cinder/Cinder/blob/glNext/src/cinder/app/cocoa/CinderViewMac.mm#L199)), I'm not entirely sure this is necessary. If anything I think we'd want to call `setNeedsDisplay:`, but I believe this already happens internally (and might be why I still get two draw calls even with the `[self draw]` commented out.

Perhaps we should put in a guard that ensures the user's draw() method (or signal) is only called once per frame.  I'm not sure at exactly what level we should do this yet, but adding it at the `ci::app::Window` level seems like less maintenance than adding it for each platform.

As far as the correctness of calling draw() twice with the same frame, it certainly isn't 'desired', though not always easy to completely avoid.

**Comment by andrewfb** [comment]
This a tricky one - the way it behaves right now is by design and in keeping with what the OS (both Mac and Windows) expect of an app - the ability to paint multiple times without returning to the message pump to accommodate live window interaction and the like. I'm open to making it an opt-out kind of thing, although a bool member var on your app could be made to prevent it as well. Open to further discussion obviously.

**Comment by richardeakin** [comment]
Hrm, it appears the draw calls are necessary - I tried limiting the draw() calls to only one per frame ([here](https://github.com/richardeakin/Cinder/commit/501737bc31bc8bbf2d958fc21005f9fff2318027)), but I see  a window full of garbage when I resize now.

Perhaps I should rephrase what I wrote above concerning correctness: in an ideal world we would only call draw() once per frame, but this isn't a very ideal world now is it. :)

Perhaps we should get to the root of why and what apps need exactly one update() per draw(). I'd prefer we find a solution for them that doesn't require us setting a flag that is needed by some blocks, but not others, and causes side effects.
